book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cumulative data members
+    - Add block to end of `self.blocks`
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cumulative data members
+    - Remove the block's hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Remove new utxos from `self.utxos`
+    - Remove the anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+
+1. Return the block
+
+**Note**: We do not subtract work from `self.partial_cummulative_work`. This
+is to make sure that the ordering of chains is stable while finalizing blocks.
+
+#### `pub fn fork(&self, new_tip: block::Hash) -> Option<Self>`
+
+Fork a chain at the block with the given hash, if it is part of this chain.
+
+1. If `self` does not contain `new_tip` return `None`
+
+2. Clone self as `forked`
+
+3. While the tip of `forked` is not equal to `new_tip`
+   - call `forked.pop_tip()` and discard the old tip
+
+4. Return `forked`
+
+#### `fn pop_tip(&mut self) -> Arc<Block>`
+
+Remove the highest height block of the non-finalized portion of a chain.
+
+1. Remove the highest height `block` from `self.blocks`
+
+2. Update cumulative data members
+    - Remove the corresponding hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Add consumed utxos and remove new utxos from `self.utxos`
+    - Remove anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+
+3. Return the block
+
+#### `Ord`
+
+The `Chain` type also implements `Ord` for reorganizing chains. First chains
+are compared by their `partial_cummulative_work`. Ties are then broken by
+comparing `block::Hash`es of the tips of each chain.
```

Comment:
@yaahc `block::Hash` does not implement `Ord`:
* in `zebra_chain::work`,  we want to compare `ExpandedDifficulty` and `block::Hash`, but reject comparisons between `block::Hash` and `block::Hash`
* in the rest of Zebra, we also want to reject comparisons between `block::Hash` and `block::Hash`

Should we make a special block::hash::ord module or something?

Reply:
> Should we make a special block::hash::ord module or something?

I don't think this would be possible, trait impls for foreign traits on local types have to be implemented in the crate that defines the type to satisfy the orphan rule. That plus it being impossible to have a private trait implementation means we couldn't implement `std::ops::Ord` on `Hash` without it being usable by everyone. We could add a `HashOrd` trait that is defined only in the state which is essentially a copy of `Ord` or even just implement it as a free function but I think theres a better option. 

In the state I implemented this by accessing the inner array of the `Hash` which does impl ord rather than adding an ord impl.

https://github.com/ZcashFoundation/zebra/pull/1069/files#diff-40f9524bc0f5eea1478a8f84f469aad3R117

Reply:
Yeah I was thinking of using the inner array. The precise order isn't actually important - it just needs to be consistent within each node.